docs/proposals/107-spark-client/README.md

```diff
@@ -0,0 +1,1023 @@
+- Enable seamless integration with Kubeflow ML pipelines and workflows
+- Provide comprehensive monitoring, logging, and debugging capabilities
+- Support production features: dynamic allocation, GPU scheduling, volumes, monitoring
+- Support interactive data exploration and notebook-style workflows via Spark Connect
+- Deliver excellent developer experience with minimal boilerplate
+
+### Non-Goals
+
+- Supporting Spark applications outside Kubernetes (local mode, standalone clusters)
+- Managing Spark cluster infrastructure (operator installation, cluster setup)
+- Replacing the Spark Operator or modifying its behavior
+- Supporting legacy Spark versions (< 3.0.0)
+- Implementing custom Spark schedulers or resource managers
+
+## Proposal
+
+We propose a **SparkClient** SDK that follows the same architectural pattern as the Kubeflow Trainer client, providing:
+
+1. **Backend Abstraction**: Pluggable backends for different deployment scenarios
+2. **Kubernetes-Native**: Direct CRD manipulation for cloud-native deployments
+3. **Unified API**: Same interface regardless of backend
+4. **Config-Based Selection**: Backend selection through typed configuration objects
+
+### Architecture Overview
+
+```
+┌─────────────────────────────────────────────────────────────┐
+│                     SparkClient (User API)                  │
+│  - submit_application(), create_session(), get_status()     │
+└──────────────────────┬──────────────────────────────────────┘
+                       │ delegates to
+                       ▼
+┌─────────────────────────────────────────────────────────────┐
+│          SparkBackend (Abstract Base Class)                 │
+│  - Defines standard methods all backends must implement     │
+└──────────────────────┬──────────────────────────────────────┘
+                       │ implemented by
+           ┌───────────┴─────────────┬──────────────────┬────────────────┐
+           ▼                         ▼                  ▼                ▼
+┌──────────────────┐   ┌──────────────────┐   ┌──────────────────┐   ┌──────────────────┐
+│OperatorBackend   │   │  GatewayBackend  │   │ ConnectBackend   │   │  LocalBackend    │
+│(Spark Operator   │   │  (REST Gateway)  │   │  (Spark Connect/ │   │  (Future)        │
+│ CRDs on K8s)     │   │                  │   │   Interactive)   │   │                  │
+└──────────────────┘   └──────────────────┘   └──────────────────┘   └──────────────────┘
+     Batch Jobs            Batch Jobs            Sessions              Local Dev
+```
+
+## Design Details
+
+### SparkClient API
+
+The main client provides a simple, intuitive API for all Spark operations:
+
+#### Basic Spark Application Submission
+
+```python
+from kubeflow.spark import SparkClient
+
+# Create client (uses Operator backend by default)
+client = SparkClient()
+
+# Submit a Spark application
+response = client.submit_application(
+    app_name="spark-pi",
```

Comment:
Can name be automatically generated similar to TrainJob and OptimizationJob by default ?

Reply:
Updated in https://github.com/kubeflow/sdk/pull/158/commits/5fc0308f7f29a3d768aa626e73052aa7a889b113